processing/src/main/java/io/druid/segment/filter/EqualColumnsFilter.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment.filter;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import io.druid.collections.bitmap.ImmutableBitmap;
+import io.druid.query.ColumnSelectorPlus;
+import io.druid.query.dimension.DefaultDimensionSpec;
+import io.druid.query.extraction.ExtractionFn;
+import io.druid.query.filter.BitmapIndexSelector;
+import io.druid.query.filter.Filter;
+import io.druid.query.filter.ValueGetter;
+import io.druid.query.filter.ValueMatcher;
+import io.druid.query.filter.ValueMatcherColumnSelectorStrategy;
+import io.druid.query.filter.ValueMatcherColumnSelectorStrategyFactory;
+import io.druid.segment.ColumnSelector;
+import io.druid.segment.ColumnSelectorFactory;
+import io.druid.segment.DimensionHandlerUtils;
+
+/**
+ */
+public class EqualColumnsFilter implements Filter
+{
+  private static final Joiner IS_JOINER = Joiner.on(" = ");
+
+  private final String[] dimensions;
+  private final ExtractionFn extractionFn;
+
+  public EqualColumnsFilter(
+      final String[] dimensions,
+      final ExtractionFn extractionFn
+  )
+  {
+    this.dimensions = Preconditions.checkNotNull(dimensions, "dimensions");
+    this.extractionFn = extractionFn;
```

Comment:
What do you think about changing this filter such that it accepts a list of DimensionSpec instead, which would allow the user to specify a separate extraction function for each dimension in the comparison? 

I can see various use cases where that functionality would be helpful (e.g., extracting and comparing a substring from two columns, but where the two column's base values have totally different formats causing a shared extractionFn to be insufficient)

The DimensionSpecs could also be directly passed to `DimensionHandlerUtils.createColumnSelectorPlus`, so that this filter doesn't need to manage the extraction functions itself in the matcher.

---

Related to that, I think it makes sense to have `EqualColumnsFilter` be a separate type of filter instead of being created from SelectorDimFilter, given that the parameters on the filter are being used differently.

The equal columns use case cares about the list of dimensions and ignores the selector value and single dimension, while the original constant value comparison case uses the value/single dimension and ignores the list of dimensions.

I think it'd be cleaner with fewer rules about what needs to be specified if the filters are totally split since the only common point now is the singular `extractionFn`.

Separation would also help remove the restriction on only having one extraction function for the equal columns case.